### ISSUE TITLE
charging classify: add a sequence-model path — the GBM variant leaves +0.012 AUROC and half the FPR on the table

### ISSUE THREAD

**Issue by jayhesselberth** [rnabioco/escapepod-rs#217]
`escpod classify` (#204, #206) consumes a **GBM** bundle. On the 20260814 LDX 16-plex — the first charging corpus where the label is not the flowcell — a small sequence model over the *same features* beats it by a margin that is stable across seeds and across a 4× change in data volume. This proposes an additive second variant; the GBM path stays as-is.

## The measurement

2.0M-read corpus (5 pod5/cell), held-out **cell** as test, 3 paired seeds, no sign flips:

| arch | AUROC | sd | balanced acc |
|---|---|---|---|
| **BiLSTM** | **0.9530** | 0.0002 | **0.9141** |
| CNN (1D over offsets) | 0.9509 | 0.0002 | 0.9103 |
| GBM (current) | 0.9408 | 0.0002 | 0.8508 |

LSTM − GBM = **+0.0122** (per-seed +0.0121 / +0.0124 / +0.0121). At one pod5 per cell it was +0.0118, so the margin does not shrink with data.

Two things matter more than the AUROC digit:

- **The false-positive rate roughly halves**, 0.10 → 0.056 on the uncharged codes (spread 0.055–0.058 across all six). For an assay whose deliverable is a charging *fraction*, that moves the prior-corrected estimate more than the ranking does.
- **It is not a shortcut.** Speed-stratified AUROC is 0.90–0.96 in every bin of body translocation speed, so it is not reading the 58/41/38 samples-per-base block signature that distinguishes the library preps.

## Why this cannot be done as GBM features instead

The obvious cheaper route is to hand the GBM the cross-offset structure as engineered columns. Measured, it does not work: within-read centred copies of every feature gain **+0.0009**. The discriminative structure is a *shape* across offsets — the residual peaks at +0/+1, inverts at +3/+4, peaks again at +12 — and `resid(+12) − resid(+3)` scores 0.7508 against 0.7299 for +12 alone. An axis-aligned tree approximates a difference badly.

The control that makes this readable: an **MLP on identical features scores 0.9364, below the GBM**. So this is not "neural beats trees" — the gain belongs specifically to models that respect the offset axis. Convolution captures most of it, recurrence a little more. (That is also the inductive bias of the original remora `ConvLSTM_w_ref` this classifier replaces.)

## Feasibility — it exports to ONNX

The reason the leech ConvLSTM was never shippable is that it only does `torch.export`. This model does not have that problem. Verified just now on the real architecture:

```
ONNX export: OK   (opset 17, dynamo=False)
max |torch - onnxruntime| = 3.35e-08
dynamic batch: (11, 2)
size: 329 KB
```

## What escpod would need

As I understand the v1 contract (escapepod-models#61), `classify` already computes the per-base features the GBM consumes — `dwell/mean/std/resid` at the declared `features.offsets`, with the k-mer table in the bundle. **The sequence model wants exactly those numbers, differently shaped.** So this is a layout + runtime change, not a new feature pipeline.

Concretely:

1. **Reshape** the existing per-offset features from a flat vector to `[batch, channels, offsets]` (channels = stats, offsets = the sequence axis). Columns are already offset-major, so this is a view.
2. **Mask channels.** Each value channel is paired with a finite/NaN indicator, doubling the channel count. The GBM handles missingness natively; a network cannot, and aliasing "unresolved" onto a plausible value is exactly the failure the raw-signal CNN suffers.
3. **Per-channel standardisation constants** (mean/sd fitted on train) carried in the bundle, applied before inference.
4. **Run the ONNX** and take `softmax(logits)[1]` as `P(charged)`, same as the GBM path.

For the current geometry that is `[B, 8, 33]` — 4 stats × (value + mask), offsets −8..+24.

Sketch, reusing the existing block names:

```json
{
  "format": "escapepod-charging-classifier/2",
  "model": {"kind": "sequence-onnx", "file": "charging_lstm.onnx", "sha256": "..."},
  "features": {"offsets": [-8, "...", 24], "stats": ["dwell","mean","std","resid"],
               "layout": "channels_offsets", "mask_channels": true,
               "standardisation": {"mean": [...], "sd": [...]}},
  "kmer_table": {"file": "levels.txt.gz", "sha256": "..."}
}
```

## Notes / open questions

- **Additive, not a replacement.** The GBM variant should keep working; a bundle declares which it is.
- **Basecaller coupling.** Worth recording on the escpod side too: this classifier is substantially a *basecall-disruption* detector — taking the `resid` expectation from the reference instead of the read's basecall costs 0.110 AUROC. escapepod-models now pins the basecall model by content sha256 + dorado version for that reason. A bundle produced against a different basecaller is a different feature space.
- Interacts with #212 (`features.resolution` / counting query bases) — the sequence model consumes whatever that produces, unchanged.
- Interacts with #210 (abstain rule) — unchanged; abstention is decided before the model runs.
- The wider feature window (`offsets` out to +24) is a separate escapepod-models change and is already contract-expressible, since `features.offsets` is declared. Offsets 17–24 are constant across the edx/ndx family; +25 is the barcode and must never be read.
- I have not read the merged `escapepod-classify` crate (it postdates my checkout), so the integration points above are inferred from the v1 contract rather than from the code.

**Comment by jayhesselberth** [comment]
Shipped in #231, released in v0.11.0. A bundle now carries `gbm` **or** `feature_model` (a small ONNX network over the same per-base features), and which arm a directory holds is a property of the bundle, never a flag — everything above the last step is shared verbatim, so the model-specific part of the pipeline is a two-armed `ChargingScorer` and nothing else.

The proposal's conclusion held: the bundle escapepod-models actually ships (`charging_feature_nn_rna004@v0.1.0`) is `arch: lstm`, hidden 96, over `level_resid@arm_le24`.

Three rules stand between the feature vector and the graph, each of which fails *silently* if guessed, so all three are declared in the bundle and checked at load: the fold from offsets-outer columns to `[channel, offset]` (verified against `features.order`, since folding the other way transposes every input and still scores), per-channel standardisation with shipped constants, and missingness as an explicit mask channel rather than `NaN`. Parity is pinned bit-exactly on reference feature vectors and against the trainer's own probabilities on a real corpus (median |Δp| 3.4e-7 over 137,660 reads).
